src/jupyterlabcontroller/services/fileserver.py

```diff
@@ -0,0 +1,422 @@
+            ),
+            V1EnvVar(name="WORBLEHAT_DIR", value="/mnt"),
+        ]
+        image = self.config.fileserver.image + ":" + self.config.fileserver.tag
+        # Specification for the user's container.
+        container = V1Container(
+            name="fileserver",
+            env=env,
+            image=image,
+            image_pull_policy=self.config.fileserver.pull_policy.value,
+            ports=[V1ContainerPort(container_port=8000, name="http")],
+            # This is a guess.  Large file transfer is quite slow at
+            # these settings, but we really don't intend to allow the RSP
+            # to be used as a bulk download service.  Sensibly-sized things
+            # such as notebooks (without enormous rendered output, anyway)
+            # are actually pretty snappy.
+            resources=V1ResourceRequirements(
+                limits={
+                    "cpu": "100m",
+                    "memory": "128M",
+                }
+            ),
+            security_context=V1SecurityContext(
+                run_as_non_root=True,
+                run_as_user=user.uid,
+                run_as_group=user.gid,
+                allow_privilege_escalation=False,
+                read_only_root_filesystem=True,
+            ),
+            volume_mounts=mounts,
+        )
+
+        # Build the pod specification itself.
+        # FIXME work out tolerations
+        #
+        podspec = V1PodSpec(
+            containers=[container],
+            restart_policy="Never",
+            security_context=V1PodSecurityContext(
+                run_as_user=user.uid,
+                run_as_group=user.gid,
+                run_as_non_root=True,
+                fs_group=user.gid,
+                supplemental_groups=[x.id for x in user.groups],
+            ),
+            volumes=volumes,
+        )
+        return podspec
+
+    def build_volumes(self) -> list[VolumeContainer]:
+        vconfig = self.config.lab.volumes
+        vols = []
+        for storage in vconfig:
+            ro = False
+            if storage.mode == FileMode.RO:
+                ro = True
+            vname = storage.container_path.replace("/", "_")[1:]
+            if not storage.server:
+                vol = V1Volume(
+                    host_path=V1HostPathVolumeSource(path=storage.server_path),
+                    name=vname,
+                )
+            else:
+                vol = V1Volume(
+                    nfs=V1NFSVolumeSource(
+                        path=storage.server_path,
+                        read_only=ro,
+                        server=storage.server,
+                    ),
+                    name=vname,
+                )
+            vm = V1VolumeMount(
+                mount_path="/mnt" + storage.container_path,
+                read_only=ro,
+                name=vname,
+            )
+            vols.append(VolumeContainer(volume=vol, volume_mount=vm))
+        return vols
```

Comment:
It's not entirely duplicated, because WebDAV can only access a single mountpoint per client, hence my aggregation of them all under `/mnt`.  I did try this with a call to the Lab Manager's build volumes and then a loop over the Volume Mounts, but it was less obvious what was happening.  I can go back and reinsert that, though.

Reply:
Moved into LabBuilder, which eventually will need a more generic name, but let's get the DM-38795 work merged first.